ghcide/test/exe/ExceptionTests.hs

```diff
@@ -0,0 +1,212 @@
+                  liftIO $ assertBool "We caught an error, but it wasn't ours!"
+                          (T.isInfixOf "divide by zero" _message && T.isInfixOf (coerce pluginId) _message)
+                _ -> liftIO $ assertFailure $ show lens
+        , testCase "Commands" $ do
+          let pluginId = "command-exception"
+              commandId = CommandId "exception"
+              plugins = pluginDescToIdePlugins $
+                  [ (defaultPluginDescriptor pluginId)
+                      { pluginCommands =
+                          [ PluginCommand commandId "Causes an exception" $ \_ (_::Int) -> do
+                              _ <- liftIO $ throwIO DivideByZero
+                              pure (InR Null)
+                          ]
+                      }]
+
+          testIde recorder (testingLite recorder logger plugins) $ do
+              _ <- createDoc "A.hs" "haskell" "module A where"
+              waitForProgressDone
+              let cmd = mkLspCommand (coerce pluginId) commandId "" (Just [A.toJSON (1::Int)])
+                  execParams = ExecuteCommandParams Nothing (cmd ^. L.command) (cmd ^. L.arguments)
+              (view L.result -> res) <- request SMethod_WorkspaceExecuteCommand execParams
+              case res of
+                Left (ResponseError {_code = InR ErrorCodes_InternalError, _message}) ->
+                  liftIO $ assertBool "We caught an error, but it wasn't ours!"
+                          (T.isInfixOf "divide by zero" _message && T.isInfixOf (coerce pluginId) _message)
+                _ -> liftIO $ assertFailure $ show res
+        , testCase "Notification Handlers" $ do
+          let pluginId = "notification-exception"
+              plugins = pluginDescToIdePlugins $
+                  [ (defaultPluginDescriptor pluginId)
+                      { pluginNotificationHandlers = mconcat
+                          [  mkPluginNotificationHandler SMethod_TextDocumentDidOpen $ \_ _ _ _ ->
+                              liftIO $ throwIO DivideByZero
+                          ]
+                        , pluginHandlers = mconcat
+                          [ mkPluginHandler SMethod_TextDocumentCodeLens $ \_ _ _-> do
+                              pure (InL [])
+                          ]
+                      }
+                    , Notifications.descriptor (cmapWithPrio LogNotifications recorder) "ghcide-core"]
+
+          testIde recorder (testingLite recorder logger plugins) $ do
+              doc <- createDoc "A.hs" "haskell" "module A where"
+              waitForProgressDone
+              (view L.result -> lens) <- request SMethod_TextDocumentCodeLens (CodeLensParams Nothing Nothing doc)
+              case lens of
+                Right (InL []) ->
+                  pure ()
+                _ -> liftIO $ assertFailure $ "We should have had an empty list" <> show lens]
+
+   , testGroup "Testing PluginError order..."
+      [ testCase "InternalError over InvalidParams" $ do
```

Comment:
I think you could abstract this a bit into something that just takes the two errors to throw, perhaps?